contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -0,0 +1,276 @@
+     *
+     * A transaction is considered valid when it hasn't expired (deadline is not met), and the signer
+     * matches the `from` parameter of the signed request.
+     *
+     * NOTE: A request may return false here but it optionally reverts in {batchExecute} to prevent
+     * reverts when a batch includes a request that was already frontrunned by another relay. This
+     * behavior is opt-in by using a flag in {executeBatch}.
+     */
+    function verify(ForwardRequestData calldata request) public view virtual returns (bool) {
+        (bool alive, bool signerMatch, ) = _validate(request, nonces(request.from));
+        return alive && signerMatch;
+    }
+
+    /**
+     * @dev Executes a `request` on behalf of `signature`'s signer guaranteeing that the forwarded call
+     * will receive the requested gas and no ETH is left stuck in the contract.
+     */
+    function execute(ForwardRequestData calldata request) public payable virtual returns (bool) {
+        // This check can be before the call because _execute reverts with an invalid request.
+        if (msg.value != request.value) {
+            revert ERC2771ForwarderMismatchedValue(request.value, msg.value);
+        }
+
+        return _execute(request, _useNonce(request.from), true);
+    }
+
+    /**
+     * @dev Batch version of {execute} with optional atomic behavior and refunding.
+     *
+     * The `atomic` parameter indicates whether all requests are executed or the whole
+     * batch reverts if a single one of them is not a valid request as defined by {verify}.
+     *
+     * The `refundReceiver` is used for optionally send the funds back to a selected address in
+     * case there's ETH left in the contract at the end of the execution instead of reverting. If
+     * the provided refund receiver is the `address(0)`, the contract will revert with an
+     * {ERC2771ForwarderMismatchedValue} error.
+     *
+     * NOTE: A refund receiver will get all contract balance back.
+     */
+    function executeBatch(
+        ForwardRequestData[] calldata requests,
+        bool atomic,
+        address payable refundReceiver
+    ) public payable virtual {
+        uint256 requestsValue;
+        uint256 spentValue;
+
+        for (uint256 i; i < requests.length; ++i) {
+            requestsValue += requests[i].value;
+            bool success = _execute(requests[i], _useNonce(requests[i].from), atomic);
+            if (success) {
+                // Will never reallistically overflow given _execute will natively revert at
+                // some point due to insufficient balance.
+                unchecked {
+                    spentValue += requests[i].value;
+                }
+            }
+        }
+
+        if (requestsValue != spentValue) {
+            if (refundReceiver == address(0)) {
+                revert ERC2771ForwarderMismatchedValue(requestsValue, msg.value);
```

Comment:
I think this is missing:

```suggestion
                revert ERC2771ForwarderMismatchedValue(requestsValue, spentValue + msg.value);
```

Reply:
No, `msg.value` is always the same during the whole call, what changes is the `address(this).balance` instead.

Reply:
There used to be a check that `requestsValue` is equal to `msg.value` but now its gone.

We should keep in mind that no this contract can still get some value trough `selfdestruct` (or its replacement) or through block reward. If that ever happens, the forwarder has a balance.

In that case its possible to send a batch that has some `requestsValue` (the fowarder balance) and no `msg.value`. If that case the call reverts, `spentValue` is 0, and the caller can "drain" the forwarder. Is that something we want ?

Otherwize we can: 
- Not track `requestsValue`
- Track `spentValue`
- if `spendValue < msg.value` then "refund" the difference (`mg.value - spendValue`).

In that case the funds send through selfdestruct and block rewards are un-recoverable.

Reply:
@ernestognw Yes, my bad.

---

It's fine IMO if funds received through selfdestruct/rewards are unrecoverable.

Reply:
> There used to be a check that requestsValue is equal to msg.value but now its gone.

It's gone because in the even of an skipped request (or many), we don't know exactly how much to refund. Instead, we need to track both consumed and requested value to refund correctly.

> We should keep in mind that no this contract can still get some value trough selfdestruct (or its replacement) or through block reward. If that ever happens, the forwarder has a balance.

The intention is not that the contract holds 0 value, but rather to make the internal accounting works even if it's overridden to hold value.

> In that case its possible to send a batch that has some requestsValue (the fowarder balance) and no msg.value. If that case the call reverts, spentValue is 0, and the caller can "drain" the forwarder. Is that something we want ?

Definitely not, but is the same as if we use `msg.value` instead. Someone can drain the contract using the same logic and that's why I decided to use this instead. Now I see it's also not viable.

> In that case the funds send through selfdestruct and block rewards are un-recoverable.

They're still recoverable (and any other balance can be stolen). Imagine the contract has 1 ether in balance and sends a batch of requests with a total value of 1 ether. If none of them executes and the relayer sends a msg.value of 2 ether, the `msg.value - spendValue` will be 1 ether, thus extracting value out of the contract.

Reply:
What is the goal here? Currently value can be extracted from the contract. 

If the goal is to refund the value that was passed and not relayed, I believe we can do that more precisely and without tracking the requested value (using the difference between what was passed through msg.value and what was passed to successful calls)

If the goal is different, then I don't get it.

Reply:
The goal is to correctly refund. Can you clarify how using `msg.value` achieved that since I already proved value can be extracted as well?